# TensorFlow Assignment: Multilayer Perceptron (MLP) and Convolutional Neural Network (CNN)

**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: Siqi Zhang

Now that you've run through a simple logistic regression model on MNIST, let's see if we can do better (Hint: we can). For this assignment, you'll build a multilayer perceptron (MLP) and a convolutional neural network (CNN), two popular types of neural networks, and compare their performance. Some potentially useful code:

In [1]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

# Import data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [2]:
# Helper functions for creating weight variables
def weight_variable(shape):
    """weight_variable generates a weight variable of a given shape."""
    initial = tf.truncated_normal(shape, stddev=0.1)
    return tf.Variable(initial)

def bias_variable(shape):
    """bias_variable generates a bias variable of a given shape."""
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

# Tensorflow Functions that might also be of interest:
# tf.nn.sigmoid()
# tf.nn.relu()

### Multilayer Perceptron

Build a multilayer perceptron for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> fully connected (500 hidden units) -> nonlinearity (Sigmoid/ReLU) -> fully connected (10 hidden units) -> softmax

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

## Test with Sigmoid

In [7]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph


### Create your MLP here##
### Make sure to name your MLP output as y_mlp ###

# W1 = tf.Variable(tf.zeros([784, 500]))
# b1 = tf.Variable(tf.zeros([500]))

W1 = weight_variable([784, 500])
b1 = bias_variable([500])
u1 = tf.matmul(x, W1) + b1

y1 = tf.nn.sigmoid(u1)

# W2 = tf.Variable(tf.zeros([500, 10]))
# b2 = tf.Variable(tf.zeros([10]))
W2 = weight_variable([500, 10])
b2 = bias_variable([10])

y_mlp = tf.matmul(y1, W2) + b2

# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_mlp))

# Optimizer
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_mlp, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(4000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(100)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy with Sigmoid: %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.059
step 250, validation accuracy 0.901
step 500, validation accuracy 0.885
step 750, validation accuracy 0.924
step 1000, validation accuracy 0.928
step 1250, validation accuracy 0.932
step 1500, validation accuracy 0.945
step 1750, validation accuracy 0.945
step 2000, validation accuracy 0.945
step 2250, validation accuracy 0.938
step 2500, validation accuracy 0.947
step 2750, validation accuracy 0.973
step 3000, validation accuracy 0.948
step 3250, validation accuracy 0.965
step 3500, validation accuracy 0.95
step 3750, validation accuracy 0.961
test accuracy 0.9636


## Test with ReLU

In [8]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph


### Create your MLP here##
### Make sure to name your MLP output as y_mlp ###

# W1 = tf.Variable(tf.zeros([784, 500]))
# b1 = tf.Variable(tf.zeros([500]))

W1 = weight_variable([784, 500])
b1 = bias_variable([500])
u1 = tf.matmul(x, W1) + b1

y1 = tf.nn.relu(u1)

# W2 = tf.Variable(tf.zeros([500, 10]))
# b2 = tf.Variable(tf.zeros([10]))
W2 = weight_variable([500, 10])
b2 = bias_variable([10])

y_mlp = tf.matmul(y1, W2) + b2

# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_mlp))

# Optimizer
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_mlp, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(4000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(100)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy with ReLU: %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.155
step 250, validation accuracy 0.946
step 500, validation accuracy 0.937
step 750, validation accuracy 0.943
step 1000, validation accuracy 0.976
step 1250, validation accuracy 0.978
step 1500, validation accuracy 0.969
step 1750, validation accuracy 0.962
step 2000, validation accuracy 0.962
step 2250, validation accuracy 0.969
step 2500, validation accuracy 0.974
step 2750, validation accuracy 0.976
step 3000, validation accuracy 0.972
step 3250, validation accuracy 0.975
step 3500, validation accuracy 0.981
step 3750, validation accuracy 0.967
test accuracy 0.9779


#### Comparison

How do the sigmoid and rectified linear unit (ReLU) compare?

***
Answer:

Based on my test on the data set, I found that the time consuming of two activation functions are both pretty short and do not differ too much.

As for the accuracy, from the two sets of output, we can find that ReLU can reach a higher accuracy, and it will reach a plateau more quickly than sigmoid (1000 steps vs. 3250 steps). 

So to some extent, for this task, ReLU performs better than sigmoid.

***

### Convolutional Neural Network

Build a simple 2-layer CNN for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> CNN (32 5x5 filters) -> nonlinearity (ReLU) ->  (2x2 max pool) -> CNN (64 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool) -> fully connected (1024 hidden units) -> nonlinearity (ReLU) -> fully connected (10 hidden units) -> softmax

Some additional functions that you might find helpful:

In [6]:
# Convolutional neural network functions
def conv2d(x, W):
    """conv2d returns a 2d convolution layer with full stride."""
    return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
    """max_pool_2x2 downsamples a feature map by 2X."""
    return tf.nn.max_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')

# Tensorflow Function that might also be of interest:
# tf.reshape()

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

*Hint: Convolutional Neural Networks are spatial models. You'll want to transform the flattened MNIST vectors into images for the CNN. Similarly, you might want to flatten it again sometime before you do a softmax. You also might want to look into the  [conv2d() documentation](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) to see what shape kernel/filter it's expecting.*

In [12]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph


### Create your CNN here##
### Make sure to name your CNN output as y_conv ###
W1 = weight_variable([5, 5, 1, 32])
b1 = bias_variable([32])

# '-1' means that the dimension here is default, which depends on the other dims.
x_reshape = tf.reshape(x, [-1, 28, 28, 1])

u1 = conv2d(x_reshape, W1) + b1
h1 = tf.nn.relu(u1)
hp1 = max_pool_2x2(h1)

W2 = weight_variable([5, 5, 32, 64])
b2 = bias_variable([64])

u2 = conv2d(hp1, W2) + b2
h2 = tf.nn.relu(u2)
hp2 = max_pool_2x2(h2)

# now the size of the image is 7*7*64

W3 = weight_variable([7 * 7 * 64, 1024])
b3 = bias_variable([1024])

hp_reshape = tf.reshape(hp2, [-1, 7 * 7 * 64])
y3 = tf.matmul(hp_reshape, W3) + b3
h3 = tf.nn.relu(y3)

W4 = weight_variable([1024, 10])
b4 = bias_variable([10])

y_conv=tf.nn.softmax(tf.matmul(h3,W4)+b4)

# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_conv))

# Optimizer
train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_conv, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(10000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(50)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.104
step 250, validation accuracy 0.604
step 500, validation accuracy 0.698
step 750, validation accuracy 0.744
step 1000, validation accuracy 0.756
step 1250, validation accuracy 0.896
step 1500, validation accuracy 0.906
step 1750, validation accuracy 0.882
step 2000, validation accuracy 0.98
step 2250, validation accuracy 0.976
step 2500, validation accuracy 0.972
step 2750, validation accuracy 0.98
step 3000, validation accuracy 0.982
step 3250, validation accuracy 0.988
step 3500, validation accuracy 0.986
step 3750, validation accuracy 0.99
step 4000, validation accuracy 0.964
step 4250, validation accuracy 0.992
step 4500, validation accuracy 0.978
step 4750, validation accuracy 0.988
step 5000, validation accuracy 0.984


KeyboardInterrupt: 

Some differences from the logistic regression model to note:

- The CNN model might take a while to train. Depending on your machine, you might expect this to take up to half an hour. If you see your validation performance start to plateau, you can kill the training.

- The logistic regression model we used previously was pretty basic, and as such, we were able to get away with using the GradientDescentOptimizer, which performs implements the gradient descent algorithm. For more difficult optimization spaces (such as the ones deep networks pose), we might want to use more sophisticated algorithms. Prof David Carlson has a lecture on this later.
    
- Because of the larger size of our network, notice that our minibatch size has shrunk.
    
- We've added a validation step every 250 minibatches. This let's us see how our model is doing during the training process, rather than sit around twiddling our thumbs and hoping for the best when training finishes. This becomes especially significant as training regimens start approaching days and weeks in length. Normally, we validate on the entire validation set, but for the sake of time we'll just stick to 10 validation minibatches (500 images) for this homework assignment.

#### Comparison

How do the MLP and CNN compare in accuracy? Training time? Why would you use one vs the other? Is there a problem you see with MLPs when applied to other image datasets?

***

Answer:

There is no doubt that MLP has a much shorter training time compared to CNN, although CNN can reach a higher accuracy (even 99%). But the tiny difference between accuracies should not compensate for the huge gap on the training time.

So for this task, I will choose MLP. To some extent, for some "small" data set, we will often prefer to using simpler models.

***